### PR TITLE
Pan 15

### PR DESCRIPTION
**Pull request by dantanxiaotian**
The unit test is not included. Still working on the unit test.

### CODE REVIEW

File: services/src/test/java/edu/msudenver/tsp/services/parser/ParserServiceTest.java [hunk 1]
```diff
@@ -18,7 +18,6 @@
 public class ParserServiceTest {
 
     @Mock private ParserService mockParserService;
-
     @InjectMocks private ParserService parserService;
 
     @Test
```
Comment (atusa17) [inline services/src/test/java/edu/msudenver/tsp/services/parser/ParserServiceTest.java]: Blocking until unit tests and integration tests are written, just to be sure no one accidentally merges this before it's done.

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 2]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+            return Optional.empty();
+        } finally {
+            LOG.info("Update user request took {} ms", Duration.between(start, Instant.now()).toMillis());
+        }
+    }
+
+
+    public boolean deleteAccount(final Account account) {
+        if(account == null){
+            LOG.error("Username does not exist, returning {}");
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: Change this error to `"Specified account is null; returning {}"`

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 3]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+        try{
+
+            final boolean persistenceApiResponse = restService.delete(persistenceApiBaseUrl + "/accounts/" + id,
+                    connectionTimeoutMilliseconds,
+                    socketTimeoutMilliseconds, HttpStatus.NO_CONTENT );
+            if(persistenceApiResponse){
+                LOG.info("return {}", persistenceApiResponse);
+            }
+            else {
+                LOG.info("Unable to delete user {}", account.toString());
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: The call to `toString()` is redundant here. Can you remove it?

File: services/src/test/java/edu/msudenver/tsp/services/UserServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import static org.mockito.Matchers.*;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class UserServiceTest {
+    @Mock private RestService restService;
+    @InjectMocks private UserService userService;
+
+    @Test
+    public void testCreateNewAccount() throws ParseException {
```
Comment (atusa17) [inline services/src/test/java/edu/msudenver/tsp/services/UserServiceTest.java]: I know this wasn't you, but can you remove the unused `throws ParseException` from all of these method signatures?

File: services/src/test/java/edu/msudenver/tsp/services/UserServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        final Account account = createAccount();
+
+        when(restService.post(anyString(), anyString(), any(TypeToken.class), anyInt(), anyInt()))
+                .thenReturn(Optional.of(account));
+
+        final Optional<Account> response = userService.createNewAccount(account);
+
+        assertTrue(response.isPresent());
+        assertEquals(account, response.get());
+    }
```
Comment (atusa17) [inline services/src/test/java/edu/msudenver/tsp/services/UserServiceTest.java]: Add a test to verify that `requestService.post()` is not called when a null account is passed and that an empty `Optional` is returned, one to verify the proper behaviour when an empty `Optional` is returned, and one to verify that an empty `Optional` is returned when an exception is thrown when trying to send the HTTP request.

File: services/src/test/java/edu/msudenver/tsp/services/UserServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        account.setId(1);
+
+        when(restService.patch(anyString(), anyString(), any(TypeToken.class), anyInt(), anyInt()))
+                .thenReturn(Optional.of(account));
+
+        final Optional<Account> response = userService.updateAccount(account);
+
+        assertTrue(response.isPresent());
+        assertEquals(account, response.get());
+    }
```
Comment (atusa17) [inline services/src/test/java/edu/msudenver/tsp/services/UserServiceTest.java]: Add a test to verify that when the passed account is null, `restService.patch()` is never called, and that an empty `Optional` is returned, one to verify that an empty `Optional` is returned when the `account.getId() == 0` and that the `restService.patch()` was never called, one for when the `persistenceApiResponse.isPresent()` is false, and one to verify that an empty `Optional` is returned when an exception is thrown inside the method.

File: services/src/test/java/edu/msudenver/tsp/services/UserServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        account.setId(1);
+
+        when(restService.delete(anyString(), anyInt(), anyInt(), any()))
+                .thenReturn(response);
+
+        final boolean persistenceApiResponse = userService.deleteAccount(account);
+
+        assertTrue(persistenceApiResponse );
+        assertEquals(response, persistenceApiResponse );
+    }
```
Comment (atusa17) [inline services/src/test/java/edu/msudenver/tsp/services/UserServiceTest.java]: Add tests to verify that when a null account is passed, `restService.delete()` is never called and the method returns `false`, one to verify that if `account.getId() == 0`, then the method returns `false` and `restService.delete()` is never called, one for when `restService.delete()` returns an empty `Optional`, and one that verifies that the method returns `false` when an exception is thrown when `restService.delete()` is called.

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 8]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+            }
+
+            return persistenceApiResponse;
+        }catch (final Exception e) {
+            LOG.error("Error deleting user {}", e);
+            return false;
+        } finally {
+            LOG.info("delete user request took {} ms", Duration.between(start, Instant.now()).toMillis());
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: I know this is knit-picky, but can you capitalize `Delete` to keep a logging convention going?

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 9]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+        final int id = account.getId();
+        final Instant start = Instant.now();
+
+        try{
+
+            final boolean persistenceApiResponse = restService.delete(persistenceApiBaseUrl + "/accounts/" + id,
+                    connectionTimeoutMilliseconds,
+                    socketTimeoutMilliseconds, HttpStatus.NO_CONTENT );
+            if(persistenceApiResponse){
+                LOG.info("return {}", persistenceApiResponse);
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: I know this is knit-picky, but can you change this to say, "Returning {}"?

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 10]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+        }
+        final int id = account.getId();
+        final Instant start = Instant.now();
+
+        try{
+
+            final boolean persistenceApiResponse = restService.delete(persistenceApiBaseUrl + "/accounts/" + id,
+                    connectionTimeoutMilliseconds,
+                    socketTimeoutMilliseconds, HttpStatus.NO_CONTENT );
+            if(persistenceApiResponse){
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: I know this is a bit knit-picky, but can you add a space between the `if` and the first `(`? Also, did you mean for this to be `persistenceApiResponse.isPresent()`? It seems as though you're only checking that the response is not null, which it always will be when it's returning an instance of `Optional`.

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 11]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+            return false;
+        }
+        final int id = account.getId();
+        final Instant start = Instant.now();
+
+        try{
+
+            final boolean persistenceApiResponse = restService.delete(persistenceApiBaseUrl + "/accounts/" + id,
+                    connectionTimeoutMilliseconds,
+                    socketTimeoutMilliseconds, HttpStatus.NO_CONTENT );
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: Delete the space at the end of this call so that it reads, `HttpStatus.NO_CONTENT);`

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 12]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+        }
+
+        if (account.getId() == 0) {
+            LOG.error("No user ID specified! Returning {}");
+            return false;
+        }
+        final int id = account.getId();
+        final Instant start = Instant.now();
+
+        try{
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: Add a space at the end of `try` so that this looks like `try {`

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 13]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+            final Optional<Account> persistenceApiResponse = restService.patch(persistenceApiBaseUrl + "accounts/" + id,
+                    new GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").create().toJson(account),
+                    typeToken,
+                    connectionTimeoutMilliseconds,
+                    socketTimeoutMilliseconds);
+
+            if (persistenceApiResponse.isPresent()) {
+                LOG.info("Returning {}", persistenceApiResponse.get());
+            } else {
+                LOG.info("Unable to update user {} with id", account.getId());
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: Change this line so that the ID is added at the end of the error:
`LOG.info("Unable to update user with id {}", account.getId());`

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 14]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
             LOG.info("Create new account request took {} ms", Duration.between(start, Instant.now()).toMillis());
         }
     }
+
+    public Optional<Account> updateAccount(final Account account) {
+        if (account == null) {
+            LOG.error("User does not exist, returning {}");
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: Change this error to be more specific: `LOG.error("Specified account is null; returning {}");`

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 15]
```diff
@@ -34,8 +35,8 @@ public UserService(final RestService restService) {
 
         try {
             final TypeToken<Account> typeToken = new TypeToken<Account>() {};
-            final Optional<Account> persistenceApiResponse = restService.post(persistenceApiBaseUrl + "/accounts/",
-                    new GsonBuilder().create().toJson(account),
+            final Optional<Account> persistenceApiResponse = restService.post(persistenceApiBaseUrl + "accounts/",
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: I know this wasn't you, but to maintain naming convention, can you rename this method to `createAccount()`?

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 16]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+
+            return persistenceApiResponse;
+        }catch (final Exception e) {
+            LOG.error("Error deleting user {}", e);
+            return false;
+        } finally {
+            LOG.info("delete user request took {} ms", Duration.between(start, Instant.now()).toMillis());
+        }
+    }
 }
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: Can you also add methods to send requests to `getListOfAccounts()`, `getAccountById()`, getAccountByUsername()` in the `AccountController`? This way, we can also send a quick query to verify whether or not a username already exists while someone is creating it, and do anything else we may need to from the `UserService`. Thank you! Don't forget to add the corresponding unit tests and integration tests for them! They should be pretty straight forward since they're just fetching data and not actually inserting or deleting any thing (CRUD operations).

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 17]
```diff
@@ -54,4 +55,76 @@ public UserService(final RestService restService) {
...
+        if (account.getId() == 0) {
+            LOG.error("No user ID specified! Returning {}");
+            return false;
+        }
+        final int id = account.getId();
+        final Instant start = Instant.now();
+
+        try{
+
+            final boolean persistenceApiResponse = restService.delete(persistenceApiBaseUrl + "/accounts/" + id,
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: The first "/"  where "/accounts/" is is redundant since there already is a "/" at the end of the `persistenceApiBaseUrl`

File: services/src/main/java/edu/msudenver/tsp/services/UserService.java [hunk 18]
```diff
@@ -25,7 +26,99 @@ public UserService(final RestService restService) {
...
+    }
+
+    public Optional<Account> getAccountById(final int id) {
+
+        if (id == 0) {
+            LOG.error("No user ID specified! Returning {}");
+            return Optional.empty();
+        }
+
+        final String auth = "";
```
Comment (atusa17) [inline services/src/main/java/edu/msudenver/tsp/services/UserService.java]: You can actually just specify the auth as null in the method since we're not using any authentication. So this variable isn't necessary. Just pass `null`
